library/core/src/intrinsics.rs

```diff
@@ -2659,7 +2660,13 @@ pub const unsafe fn const_deallocate(_ptr: *mut u8, _size: usize, _align: usize)
 macro_rules! assert_unsafe_precondition {
     ($message:expr, ($($name:ident:$ty:ty = $arg:expr),*$(,)?) => $e:expr $(,)?) => {
         {
-            #[inline(never)]
+            // When the standard library is compiled with debug assertions, we want the check to inline for better performance.
+            // This is important when working on the compiler, which is compiled with debug assertions locally.
+            // When not compiled with debug assertions (so the precompiled std) we outline the check to minimize the compile
+            // time impact when debug assertions are disabled.
+            // It is not clear whether that is the best solution, see #120848.
+            #[cfg_attr(debug_assertions, inline(always))]
+            #[cfg_attr(not(debug_assertions), inline(never))]
```

Comment:
This checks `cfg(debug_assertions)` which is different from `intrinsics::debug_assertions`. Shouldn't the comment explain why that mismatch is okay? It is certainly confusing to me.

Reply:
Why wouldn't the difference be okay? Sure, the behavior isn't exactly the same, but users who enable that will get more checks in case they don't enable debug_assertions downstream but that seems okay. We should be able to remove this again after #121114 anyways, I mostly PRed this to avoid thinking in #121114, but I have done the thinking.

Reply:
> Why wouldn't the difference be okay?

There's a funny case to consider where debug assertions are disabled but `intrinsics::debug_assertions` returns true. I would have expected the comment to acknowledge this and explain why what happens in that case is what we want.

> We should be able to remove this again after #121114 anyways

Ah, if this is temporary then never mind. :)

Reply:
> There's a funny case to consider where debug assertions are disabled but intrinsics::debug_assertions returns true.

when building `core`, that cannot happen. When building user code, yes, that's actually the expected and normal path, core without debug assertions, but debug assertions in the user case.